Improve `wxFileConfig` XDG compliance

vadz opened the pull request:
The main goal here is to allow easily creating config files in `~/.config/appname` instead of using ~/.appname` directory for them as `wxCONFIG_USE_SUBDIR` does now. Unfortunately achieving this is not that simple without breaking backwards compatibility -- which is something we really can't do here as this would break all the existing applications that wouldn't find their config files any longer.

So add a new `wxCONFIG_USE_XDG` flag which can be used to explicitly opt-in the new behaviour.

Also add `Dir_Config` which can be used to retrieve this directory location for code not using `wxFileConfig` too.

Also includes some other related cleanup.

vadz committed:
Remove unnecessary wxUnusedVar() from wxFileConfig code

No real changes, just don't use wxUnusedVar() for a parameter which is
actually used.

vadz committed:
Document that GetUserDataDir() doesn't respect XDG file layout

And recommend using GetUserDir() instead.

vadz committed:
Make wxStandardPaths::AppendAppInfo() public

This function can be useful to the application code and will be used by
wxFileConfig in the upcoming commits.

vadz committed:
Add wxStandardPaths::Dir_Config

This allows using GetUserDir(Dir_Config) to retrieve XDG-compliant
location of the configuration files, independently of the file layout
used by the global wxStandardPaths object.

Recommend using this directory instead of GetUserDataDir() for multiple
configuration files under Unix.

vadz committed:
Add [[nodiscard]] attribute to wxStandardPaths::AppendAppInfo()

It doesn't make sense to call this function and ignore its return value,
but it may happen accidentally as it might seem that it modifies its
parameter.

vadz committed:
Add wxCONFIG_USE_XDG to use with wxFileConfig

Using this style allows to tell wxFileConfig to use XDG-compliant
location for the user configuration file, even when not using XDG file
layout for wxStandardPaths, which can be convenient when modifying the
existing application using wxStandardPaths for other files locations
too.

And using it in combination with wxCONFIG_USE_SUBDIR allows to put the
config file in a XDG-compliant subdirectory, which wasn't easily
possible at all before.

vadz committed:
Use const references to wxStandardPaths in wxFileConfig code

No real changes, just minor cleanup.

dconnet commented on interface/wx/stdpaths.h:
```diff
@@ -79,6 +79,18 @@ class wxStandardPaths
+
+            Example return values:
+            - Unix: `~/.config`
+            - Windows: `C:\Users\username\AppData\Roaming`
```
Should there be a Local version of this (for Windows) like there is for userdatadir? (to allow windows users to keep this in appdata\local)

dconnet replied:
Update: Though looking at wxFileConfig closer, I notice it does not support Local, so this may not be too important.

vadz replied:
`GetUserDir(Dir_Config)` returns the same directory as `GetUserConfigDir()` under Windows as I think this makes most sense.

vadz committed:
Fix crash on shutdown if wxConfig couldn't be saved

If saving wxConfig to the file in wxAppConsoleBase::CleanUp() failed,
wxLog messages were logged into a wxLogGui target which was recreated
because wxTheApp is still non-null when CleanUp() is called. However it
is destroyed soon afterwards and showing messages from wxLogGui without
wxTheApp leads to crashes, at least in wxGTK.

To avoid this, we could set wxTheApp to null before calling CleanUp(),
but it's not clear if there is no cleanup code (or something called from
it) which doesn't rely on wxTheApp existence, so instead just flush, and
reset, the log target after calling CleanUp() but before destroying the
application object.

This fixes the crash which could be easily reproduced simply by making
the file used by wxFileConfig not-writeable under Linux.

vadz committed:
Add wxHAS_CONFIG_AS_{REG,FILE}CONFIG symbols

Sometimes it's useful to have some code used only if wxConfig is defined
as wxRegConfig or only if it is defined as wxFileConfig and testing for
these symbols is more clear than testing for the platform (and more
correct, considering that setting wxUSE_CONFIG_NATIVE to 0 may result in
wxFileConfig being used even under Windows).

vadz committed:
Add wxFileConfig::MigrateLocalFile()

This can be useful for the existing applications switching to using
XDG-compliant config files location as they can just call this function
on startup to keep using the existing file.

vadz commented:
I've added a convenient (IMO) function for migrating existing config files to the new location, please let me know if anybody has any comments about this.

I've also fixed a crash in wxGTK, hopefully without breaking anything else. I plan to cherry pick the commit doing this ([055c4cb](https://github.com/wxWidgets/wxWidgets/pull/24180/commits/055c4cbed5006f801d6e01a88eea46d89a67dc39)) to 3.2 too.

vadz commented:
I'd like to merge this soon, as I have other changes depending on this one, please let me know if there are any comments.
